File: sort_efficient/sort_efficient__merge.cs
// 実行時間 580ms
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static int o = 0;

    static void Main()
    {
        int n = int.Parse( ReadLine() );
        int[] a = ReadLine().Split().Select( int.Parse ).ToArray();

        MergeSort( a, 0, n );
        WriteLine( string.Join( " ", a ) );
        WriteLine( o );
        return;
    }

    static void MergeSort( int[] a, int l, int r )
    {
        if( l + 1 < r ) {
            int m = ( l + r ) / 2;

            MergeSort( a, l, m );
            MergeSort( a, m, r );
            Merge( a, l, m, r );
        }
        return;
    }

    static void Merge( int[] a, int l, int m, int r )
    {
        int li = 0;
        int ri = 0;
        int nl = m - l;
        int nr = r - m;
        int[] al = new int[nl + 1];
        int[] ar = new int[nr + 1];

        foreach( int i in Range( 0, nl ) ) {
            al[i] = a[l + i];
        }
        foreach( int i in Range( 0, nr ) ) {
            ar[i] = a[m + i];
        }
        al[nl] = int.MaxValue;
        ar[nr] = int.MaxValue;
        for( int i = l; i < r; i++ ) {
            if( al[li] < ar[ri] ) {
                a[i] = al[li];
                li++;
            } else {
                a[i] = ar[ri];
                ri++;
                o++;
            }
        }
        return;
    }
}

File: sort_efficient/sort_efficient__merge.py
# 実行時間 2630ms
def merge(a, l, m, r):
    global o
    li = 0
    ri = 0
    nl = m - l
    nr = r - m
    al = [0] * (nl + 1)
    ar = [0] * (nr + 1)

    for i in range(nl):
        al[i] = a[l + i]
    for i in range(nr):
        ar[i] = a[m + i]
    al[nl] = pow(10, 10) * 2
    ar[nr] = pow(10, 10) * 2
    for i in range(l, r):
        if al[li] < ar[ri]:
            a[i] = al[li]
            li += 1
        else:
            a[i] = ar[ri]
            ri += 1
            o += 1

def mergeSort(a, l, r):
    if l + 1 < r:
        m = (l + r) // 2

        mergeSort(a, l, m)
        mergeSort(a, m, r)
        merge(a, l, m, r)

n = int(input())
o = 0
a = list(map(int, input().split()))

mergeSort(a, 0, n)
print(*a)
print(o)
